/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2025 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : It is required to call "systemctl start mdns"
*                         imx_otbr_gpiochip_spidev_cfg.sh Shell script allows to modify the /etc/default/otbr-agent
*                         Modification consists in replacing specific string patterns by using the DTB spidev and gpio naming
*                         Consequently, SPI OTBR_AGENT_OPTS of /etc/default/otbr-agent contains the right spidev number to use
*                         and the right gpiochip number and line
*                         otbr-agent.service is updated accordingly
* Version               : v001
* Mandatory/Optional    : Mandatory
* Impact if not applied : /etc/default/otbr-agent file can't beneficiate of DTB gpio naming.
*/

Upstream-Status: Inappropriate  [platform specific changes]

diff --git a/src/agent/CMakeLists.txt b/src/agent/CMakeLists.txt
index 7060df29b47..4c27f41298f 100644
--- a/src/agent/CMakeLists.txt
+++ b/src/agent/CMakeLists.txt
@@ -69,7 +69,7 @@ set(OTBR_AGENT_USER "root" CACHE STRING "set the username running otbr-agent ser
 set(OTBR_AGENT_GROUP "root" CACHE STRING "set the group using otbr-agent client")
 
 if(OTBR_MDNS STREQUAL "mDNSResponder")
-    set(EXEC_START_PRE "ExecStartPre=/usr/sbin/service mdns start\n")
+    set(EXEC_START_PRE "ExecStartPre=systemctl start mdns\nExecStartPre=/usr/sbin/imx_otbr_gpiochip_spidev_cfg.sh\n")
 elseif(OTBR_MDNS STREQUAL "avahi")
     set(EXEC_START_PRE "ExecStartPre=/usr/sbin/service avahi-daemon start\n")
 else()
@@ -104,6 +104,11 @@ install(FILES ${CMAKE_CURRENT_BINARY_DIR}/otbr-agent.default
     RENAME otbr-agent
 )
 
+configure_file(imx_otbr_gpiochip_spidev_cfg.sh.in imx_otbr_gpiochip_spidev_cfg.sh)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/imx_otbr_gpiochip_spidev_cfg.sh
+    DESTINATION ${OTBR_SYSTEMD_UNIT_DIR}
+)
+
 pkg_check_modules(LIBSYSTEMD libsystemd)
 if(LIBSYSTEMD_FOUND)
     target_compile_definitions(otbr-config INTERFACE "HAVE_LIBSYSTEMD=1")
diff --git a/src/agent/imx_otbr_gpiochip_spidev_cfg.sh.in b/src/agent/imx_otbr_gpiochip_spidev_cfg.sh.in
new file mode 100644
index 00000000000..88809e4c804
--- /dev/null
+++ b/src/agent/imx_otbr_gpiochip_spidev_cfg.sh.in
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+echo "From device-tree settings -> Retrieve spidev numbers, IW61x_OT_RST and IW61w_OT_INT gpiochip lines"
+
+RST_gpiochip_num=`gpioinfo IWxxx_NB_IND_RST_15_4 2>/dev/null | awk -F" " '{print $1}'`
+RST_gpiochip_line=`gpioinfo IWxxx_NB_IND_RST_15_4 2>/dev/null | awk -F" " '{print $2}'`
+INT_gpiochip_num=`gpioinfo IWxxx_NB_SPI_INT 2>/dev/null | awk -F" " '{print $1}'`
+INT_gpiochip_line=`gpioinfo IWxxx_NB_SPI_INT 2>/dev/null | awk -F" " '{print $2}'`
+
+spi0=`find /sys/firmware/devicetree/base -name spi@0`
+if [ -f $spi0/label ] && [ `tr -d '\0' < $spi0/label` = "IWxxx_SPIDEV" ]; then
+    num=$(echo $spi0 | awk -F/ '{print NF}')
+    let num=num-1
+    spi_bus_addr=$(echo $spi0 | cut -d '/' -f$num | cut -d "@" -f2)
+    spi_folder=$(find /sys/devices/platform -type d -name $spi_bus_addr.spi*)
+    SPI_dev=$(find $spi_folder -type d -name spidev*.* | awk -F/ '{print $NF}')
+fi
+
+echo "SPI device:   $SPI_dev"
+echo "RST:          $RST_gpiochip_num:$RST_gpiochip_line"
+echo "INT:          $INT_gpiochip_num:$INT_gpiochip_line"
+
+export SPI_dev=$SPI_dev
+export RST_gpiochip_num=$RST_gpiochip_num
+export RST_gpiochip_line=$RST_gpiochip_line
+export INT_gpiochip_num=$INT_gpiochip_num
+export INT_gpiochip_line=$INT_gpiochip_line
+
+sed -i '1 s/.*/#WARNING, this file was modified by '"$(basename $0)"'/' /etc/default/otbr-agent
+sed -i 's/spidev_pattern/'$SPI_dev'/g' /etc/default/otbr-agent
+sed -i 's/RST_gpiochip_num_pattern/'$RST_gpiochip_num'/g' /etc/default/otbr-agent
+sed -i 's/RST_gpiochip_line_pattern/'$RST_gpiochip_line'/g' /etc/default/otbr-agent
+sed -i 's/INT_gpiochip_num_pattern/'$INT_gpiochip_num'/g' /etc/default/otbr-agent
+sed -i 's/INT_gpiochip_line_pattern/'$INT_gpiochip_line'/g' /etc/default/otbr-agent
+sed -i 's/extra_opts //g' /etc/default/otbr-agent
+

diff --git a/src/agent/otbr-agent.default.in b/src/agent/otbr-agent.default.in
index f5cdaf94652..3d78ae73315 100644
--- a/src/agent/otbr-agent.default.in
+++ b/src/agent/otbr-agent.default.in
@@ -1,5 +1,10 @@
 # Default settings for otbr-agent. This file is sourced by systemd
 
 # Options to pass to otbr-agent
-OTBR_AGENT_OPTS="-I wpan0 -B @OTBR_INFRA_IF_NAME@ @OTBR_RADIO_URL@ trel://@OTBR_INFRA_IF_NAME@"
+#UART
+#OTBR_AGENT_OPTS="-d 1 -I wpan0 -B eth0 'spinel+hdlc+uart:///tmp/ttyOpenThread extra_opts ' trel://eth0"
+
+#SPI
+OTBR_AGENT_OPTS="-d 1 -I wpan0 -B eth0 'spinel+spi:///dev/spidev_pattern?gpio-reset-device=/dev/RST_gpiochip_num_pattern&gpio-int-device=/dev/INT_gpiochip_num_pattern&gpio-int-line=INT_gpiochip_line_pattern&gpio-reset-line=RST_gpiochip_line_pattern&spi-mode=0&spi-speed=2000000&spi-reset-delay=0 extra_opts ' trel://eth0"
+
 OTBR_NO_AUTO_ATTACH=@OTBR_NO_AUTO_ATTACH@
diff --git a/src/agent/otbr-agent.service.in b/src/agent/otbr-agent.service.in
index 8314121347f..ec687009279 100644
--- a/src/agent/otbr-agent.service.in
+++ b/src/agent/otbr-agent.service.in
@@ -1,12 +1,12 @@
 [Unit]
 Description=OpenThread Border Router Agent
-ConditionPathExists=@CMAKE_INSTALL_FULL_SBINDIR@/otbr-agent
+ConditionPathExists=/usr/sbin/otbr-agent
 Requires=dbus.socket
 After=dbus.socket
 
 [Service]
-EnvironmentFile=-@CMAKE_INSTALL_FULL_SYSCONFDIR@/default/otbr-agent
-@EXEC_START_PRE@ExecStart=@CMAKE_INSTALL_FULL_SBINDIR@/otbr-agent $OTBR_AGENT_OPTS
+EnvironmentFile=-/etc/default/otbr-agent
+@EXEC_START_PRE@ExecStart=/usr/sbin/otbr-agent $OTBR_AGENT_OPTS
 KillMode=mixed
 Restart=on-failure
 RestartSec=5
-- 
2.43.0

